packages/env/lib/config/post-process-config.js

```diff
@@ -0,0 +1,195 @@
+const { ValidationError } = require( './validate-config' );
+
+/**
+ * @typedef {import('./config').WPServiceConfig} WPServiceConfig
+ */
+
+/**
+ * Performs any additional post-processing on the config object.
+ *
+ * @param {WPServiceConfig} config The config to process.
+ *
+ * @return {WPServiceConfig} The config after post-processing.
+ */
+module.exports = function postProcessConfig( config ) {
+	// Make sure that we're operating on a config object that has
+	// complete environment configs for convenience.
+	config = mergeRootToEnvironments( config );
+
+	config = appendPortToWPConfigs( config );
+
+	validate( config );
+	return config;
+};
+
+/**
+ * Merges the root config and each environment together in order to make sure each environment has
+ * a full config object to work with internally. This makes it easier than having to try and
+ * resolve the full config options every time we want to use them for something.
+ *
+ * @param {WPServiceConfig} config The config to process.
+ *
+ * @return {WPServiceConfig} The config object with the root options merged together with the environment-specific options.
+ */
+function mergeRootToEnvironments( config ) {
+	// Some root-level options need to be merged early because they have a special
+	// cascade behavior that would break the normal merge. After merging we then
+	// delete them to avoid that breakage and add them back before we return.
+	const removedRootOptions = {};
+	if (
+		config.port !== undefined &&
+		config.env.development.port === undefined
+	) {
+		removedRootOptions.port = config.port;
+		config.env.development.port = config.port;
+		delete config.port;
+	}
+	if (
+		config.testsPort !== undefined &&
+		config.env.tests.port === undefined
+	) {
+		removedRootOptions.testsPort = config.testsPort;
+		config.env.tests.port = config.testsPort;
+		delete config.testsPort;
+	}
+
+	// Merge the root config and the environment configs together so that
+	// we can ignore the root config and have full environment configs.
+	for ( const env in config.env ) {
+		config.env[ env ] = mergeConfigs(
+			deepCopyRootOptions( config ),
+			config.env[ env ]
+		);
+	}
+
+	// Set any root-level options we reset back.
+	for ( const option in removedRootOptions ) {
+		config[ option ] = removedRootOptions[ option ];
+	}
+
+	return config;
+}
+
+/**
+ * Appends the configured port to certain wp-config options.
+ *
+ * @param {WPServiceConfig} config The config to process.
+ *
+ * @return {WPServiceConfig} The config after post-processing.
+ */
+function appendPortToWPConfigs( config ) {
+	const options = [ 'WP_TESTS_DOMAIN', 'WP_SITEURL', 'WP_HOME' ];
+
+	// We are only interested in editing the config options for environment-specific configs.
+	// If we made this change to the root config it would cause problems since they would
+	// be mapped to all environments even though the ports will be different.
+	for ( const env in config.env ) {
```

Comment:
Since we're accessing `config.env[env]` so much, I might use `for (const [_, env] of Object.entries(config.env))` instead, and then `env` is just the total env object

Reply:
Honestly, I kind of like things like this to be a little bit more expressive about what exactly you're editing. I can change it if you really want me to though :smile:

Reply:
I do personally feel like it's more readable and a bit less noisy with object.entries, but I don't care that much :)